### ISSUE TITLE
Detection of duplicate enum literals in dictionary keys (F601)

### ISSUE THREAD

**Issue by tsx** [astral-sh/ruff#13662]
[multi-value-repeated-key-literal (F601)](https://docs.astral.sh/ruff/rules/multi-value-repeated-key-literal/#multi-value-repeated-key-literal-f601) only works with primitive literals like keys and strings. In our codebase we use enums extensively for increased type safety over strings, and that includes mappings of enum values to other values. I'd love ruff to find issues in those too.

This triggers F601:
```python
string_keys = {"a": 1, "a": 2}
```

This doesn't trigger but should:
```python
from enum import Enum
class E(Enum):
    A = "a"

enum_keys = {E.A: 1, E.A: 2}
```

(ruff 0.6.9)

Thanks for the awesome tool!

**Comment by MichaReiser** [comment]
That makes sense but isn't something that ruff can support today without proper multifile analysis (to support enums imported from other files) and type inference (to get the types of enum constants).

**Comment by tsx** [comment]
Would it make sense to make a separate rule that checks for identical expressions in keys? It can have false positives if the expression has side effects and changes value on every evaluation, but one can argue that using those as "keys" in a dictionary is very confusing and obscures the code intent.

**Comment by MichaReiser** [comment]
Good point. We could extend the rule to support identical attribute chains.

**Comment by robin-atlasbio** [comment]
I just wanted to open an issue about F602 but thankfully found this one. 

```py
from enum import StrEnum

FOO = "foo"


class SomeConstants:
    FOO = "foo"
    BAR = "bar"


class SomeEnum(StrEnum):
    FOO = "foo"
    BAR = "bar"


x = {
    FOO: 1,
    FOO: 2,  # Ruff(F602)
    SomeConstants.FOO: 1,
    SomeConstants.FOO: 2,  # No error!
    SomeEnum.FOO: 1,
    SomeEnum.FOO: 2,  # No Error
}
```

In our case we use the class constant collection pattern with nested classes:

```py
class SomeCategory:
    class SomeChild:
        class Inner:
            FOO = "foo"
            BAR = "bar"
```

Which can make it difficult to see this mistake - meaning it would be great if this was implemented. I'm not sure if there is any case where a same-expression could legitimately appear multiple times in dict keys - even if technically it could yield different values.


Even a function call I think would almost always be a mistake:

```py

class NominalStr(str):
  __slots__ = ()


d = {
  NominalStr("hi"): 1,
  NominalStr("hi"): 2
}
```

Even though I'd see how this one _might_ in some rare case be intentional.
